### ISSUE TITLE
[no-unnecessary-condition] no-unnecessary-condition doesn't seem to permit type guards

### ISSUE THREAD

**Issue by metaquanta** [typescript-eslint/typescript-eslint#2478]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

```TS
export function isCallable<T, V>(f: ((p: V) => T) | T): f is (p: V) => T {
  return (f as (p: V) => T).call !== undefined;
}
```
My [tsconfig](https://github.com/metaquanta/tilings/blob/tslint/tsconfig.json)

**Expected Result**
I expect some way to define a custom type guard that keeps `no-unnecessary-condition` happy.

Is there some other form that works? I can't figure out how to do overloads without type-guards.

**Actual Result**
`error  Unnecessary conditional, the types have no overlap  @typescript-eslint/no-unnecessary-condition`

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.10.1` |
| `@typescript-eslint/parser`        | `3.10.1` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.7.0` |
| `node`                             | `14.9.0` |

**Comment by bradzacher** [comment]
The rule is driven off of the types.

`.call` always exists on a function type. It's never undefined - hence the rule errors.

If you are intentionally writing a condition that's unnecessary so they you can do some strict runtime validation, you should either change your types so that the check looks necessary, or you should use a disable comment.

**Comment by metaquanta** [comment]
To write a type guard with `no-unnecessary-condition`, cast to an anonymous type where the checked property is optional. For instance:

```
 export function isCallable<T, V>(f: ((p: V) => T) | T): f is (p: V) => T {
  return (f as { call?: unknown }).call !== undefined;
}
```
